module4/persistence/src/main/java/com/epam/esm/persistence/impl/TagDaoImpl.java

```diff
@@ -0,0 +1,73 @@
+package com.epam.esm.persistence.impl;
+
+import com.epam.esm.criteria_info.TagCriteriaInfo;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.persistence.TagDao;
+import com.epam.esm.persistence.dataspecification.TagSpecification;
+import com.epam.esm.persistence.repository.TagRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Repository;
+
+import javax.persistence.EntityManager;
+import java.math.BigInteger;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+@Repository
+@RequiredArgsConstructor
+public class TagDaoImpl implements TagDao {
```

Comment:
@ViktoriaShamuradova I belive that it should executes on service layer. For example you canprepare builder for specifications. After that from service invouke repository layer.This implementation is very confusing.

Reply:
If I understand correctly you offer to have small specifications defined in reposiroty level and build them in service (please correct me if I'm wrong). I also agree with that approach and it was done like this previously. But I was asked to move all the specifications to the repository level and just pass a `criteria info` (something like search request) object to it